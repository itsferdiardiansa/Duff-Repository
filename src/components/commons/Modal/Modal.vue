<template>
  <div class="fixed z-50 inset-0 overflow-y-auto" v-if="isShow">
    <div class="flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0">
      <div class="fixed inset-0 transition-opacity" aria-hidden="true">
        <div class="absolute inset-0 bg-gray-500 opacity-75"></div>
      </div>

      <span class="hidden sm:inline-block sm:align-middle sm:h-screen">&#8203;</span>
      
      <div class="modal-body" role="dialog" aria-modal="true" aria-labelledby="modal-headline">
        <div class="bg-white px-4 pt-5 pb-4 sm:p-6 sm:pb-4">
          <div class="sm:flex sm:items-start">
            <div class="mt-3 text-center sm:mt-0 sm:ml-4 sm:text-left">
              
              <h3 class="text-lg leading-6 font-medium text-gray-900" id="modal-headline">
                <slot name="header"></slot>
              </h3>
              
              <div class="mt-2">
                <slot name="body"></slot>
              </div>
            </div>
          </div>
        </div>

        <div class="bg-gray-50 px-4 py-3 sm:px-6 sm:flex sm:flex-row-reverse" v-if="withConfirmation">
          <button type="button" class="btn-confirm">
            Yes
          </button>
          <button 
            type="button" 
            class="btn-cancel"
            @click="toggleModal"
          >
            No
          </button>
        </div>
      </div>
    </div>
  </div>
</template>
<script>
import { ref } from 'vue'

export default {
  props: {
    withConfirmation: {
      type: Boolean,
      default: true
    },
    onConfirm: {
      type: Function,
      default: () => { }
    }
  },
  setup(props) {
    let isShow = ref(true)

    const toggleModal = () => {
      isShow.value = !isShow.value
    }

    const handleConfirm = () => {
      props.onConfirm()
    }

    return {
      isShow,
      toggleModal,
      handleConfirm
    }
  }
}
</script>
<style lang="scss" scoped>
.modal-body {
  @apply inline-block align-bottom bg-white rounded-lg text-left; 
  @apply overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-lg sm:w-full;

  .btn-confirm {
    @apply w-full inline-flex justify-center rounded-md border border-transparent shadow-sm px-4 py-2 bg-red-600 text-base font-medium text-white;
    @apply hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-red-500 sm:ml-3 sm:w-auto sm:text-sm;
  }

  .btn-cancel {
    @apply mt-3 w-full inline-flex justify-center rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-base font-medium text-gray-700;
    @apply hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm;
  }
}
</style>