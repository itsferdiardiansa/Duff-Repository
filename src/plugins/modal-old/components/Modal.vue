<template>
  <!-- <teleport to="#ss-modal"> -->
  <div :class="`${prefixClass}-modal--body`" ref="modalEl">
    <div :class="`${prefixClass}-modal--panel`">
      <div :class="`${prefixClass}-modal--panel-content`">
        <h3 class="content-header" v-if="title || $slots.header">
          <slot name="header">
            {{ title }}
          </slot>
        </h3>

        <div class="content-desc">
          <component :is="getContent" />
        </div>
      </div>
    </div>

    <div ref="modalFooterEl" :class="getFooterClass" v-if="footer">
      <component :is="getFooter" />
    </div>
  </div>
  <!-- </teleport> -->
</template>
<script>
/* eslint-disable */
import {
  computed,
  getCurrentInstance,
  h,
  inject,
  nextTick,
  onMounted,
  ref,
} from 'vue';
import Button from '@common/Button';
import ModalDefaultFooter from './ModalFooter';

export default {
  name: 'PLModal',
  components: {
    Button,
  },
  props: {
    name: {
      type: String,
    },
    title: {
      type: String,
      default: '',
    },
    content: {
      type: [String, Object, Function],
      default: '',
    },
    footer: {
      type: [Boolean, Object, Function],
      default: ModalDefaultFooter,
    },
    footerAlign: {
      type: String,
      default: '',
    },
    withConfirmation: {
      type: Boolean,
      default: true,
    },
    onConfirmFn: {
      type: Function,
      default: () => {},
    },
    onCancelFn: {
      type: Function,
      default: () => {},
    },
    afterClose: {
      type: Function,
      default: () => {},
    },
    closeable: {
      type: Boolean,
      default: true,
    },
  },
  emits: ['onConfirm', 'onCancel'],
  setup(props, { attrs, slots, emit }) {
    const modalEl = ref();
    const modalBodyEl = ref();
    const modalFooterEl = ref();
    const modalConfirmBtnEl = ref();
    const modalCancelBtnEl = ref();
    const modalContext = inject('modalContext');
    const root = getCurrentInstance();

    const handleConfirm = () => {
      toggleModal();
      props.onConfirmFn(hideModal);
    };

    const handleCancel = () => {
      hideModal();
      props.onCancelFn();
    };

    const toggleModal = () => {
      if (props.closeable) hideModal();
    };

    const hideModal = () => {
      setTimeout(() => {
        modalContext.emitter.emit('hide-modal', props);
      }, 100);
    };

    const getContent = computed(() => {
      const { content } = props;
      let childContent = {};

      if (typeof content === 'string' && content) {
        childContent = h('label', {}, content);

        return childContent;
      }
      return content ? content : slots.default;
    });

    const getFooter = computed(() => {
      const { footer } = props;

      if (slots.footer) return slots.footer;

      if (
        typeof footer !== 'boolean' &&
        footer.name === 'PLModalFooterDefault'
      ) {
        return h(footer, { handleConfirm, handleCancel });
      }

      return h(footer, { hideModal });
    });

    const getFooterClass = computed(() => {
      const {
        data: { prefixClass },
      } = getCurrentInstance();
      const { footerAlign } = props;
      let footerClass = [`${prefixClass}-modal--footer`];

      if (footerAlign) footerClass.push(`align-${footerAlign}`);

      return footerClass;
    });

    onMounted(() => {
      if (
        Reflect.has(modalContext, 'pushModal') &&
        modalContext !== undefined
      ) {
        modalContext.pushModal(root.vnode, modalEl);
      }
    });

    return {
      modalEl,
      modalBodyEl,
      modalFooterEl,
      modalConfirmBtnEl,
      modalCancelBtnEl,
      hideModal,
      handleConfirm,
      handleCancel,
      toggleModal,
      getContent,
      getFooter,
      getFooterClass,
    };
  },
};
</script>
<style lang="scss" src="./styles.scss" scoped></style>
