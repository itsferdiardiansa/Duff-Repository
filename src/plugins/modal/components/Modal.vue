<template>
  <template v-if="isShow">
    <div :class="`${prefixClass}-modal`" ref="modalEl">
      <div :class="`${prefixClass}-modal--container`">
        <div
          :class="`${prefixClass}-modal--overlay`"
          aria-hidden="true"
          @click="toggleModal(false)"
        >
          <div :class="`${prefixClass}-modal--overlay-bg`"></div>
        </div>

        <span class="hidden sm:inline-block sm:align-middle sm:h-screen"
          >&#8203;</span
        >

        <div :class="`${prefixClass}-modal--body`" ref="modalBodyEl">
          <div :class="`${prefixClass}-modal--panel`">
            <div :class="`${prefixClass}-modal--panel-content`">
              <h3 class="content-header" v-if="title">
                <slot name="header">
                  {{ title }}
                </slot>
              </h3>

              <div class="content-desc">
                <component :is="getContent" />
              </div>
            </div>
          </div>

          <div ref="modalFooterEl" :class="getFooterClass" v-if="footer">
            <component :is="getFooter" />
          </div>
        </div>
      </div>
    </div>
  </template>
</template>
<script>
/* eslint-disable */
import { computed, getCurrentInstance, h, inject, onMounted, ref } from 'vue';
import Button from '@common/Button';
import ModalDefaultFooter from './ModalFooter';

export default {
  name: 'PLModal',
  components: {
    Button,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    content: {
      type: [String, Object],
      default: '',
    },
    footer: {
      type: [Boolean, Object],
      default: ModalDefaultFooter,
    },
    footerAlign: {
      type: String,
      default: '',
    },
    withConfirmation: {
      type: Boolean,
      default: true,
    },
    onConfirmFn: {
      type: Function,
      default: () => {},
    },
    onCancelFn: {
      type: Function,
      default: () => {},
    },
    onShow: {
      type: Boolean,
      default: false,
    },
  },
  created() {},
  setup(props, { attrs, slots }) {
    const modalEl = ref();
    const modalBodyEl = ref();
    const modalFooterEl = ref();
    const modalConfirmBtnEl = ref();
    const modalCancelBtnEl = ref();
    const isShow = ref(false);
    const modalContext = inject('modalContext');
    const root = getCurrentInstance();
    let currentData;

    const handleConfirm = () => {
      toggleModal(false);
      props.onConfirmFn(currentData);
    };

    const handleCancel = () => {
      toggleModal(false);
      props.onCancelFn();
    };

    const toggleModal = status => {
      isShow.value = status;

      setTimeout(() => {
        modalContext.emitter.emit('hide-modal');
      }, 100);
    };

    const modal = {
      show: data => {
        isShow.value = true;
        currentData = data;
      },
    };

    const getContent = computed(() => {
      const { content } = props;
      let childContent = {};

      if (typeof content === 'string' && content) {
        childContent = h('label', {}, content);

        return childContent;
      }

      return content ? content : slots.default;
    });

    const getFooter = computed(() => {
      const { footer } = props;

      if (
        typeof footer !== 'boolean' &&
        footer.name === 'PLModalFooterDefault'
      ) {
        return h(footer, { handleConfirm, handleCancel });
      }

      return footer;
    });

    const getFooterClass = computed(() => {
      const {
        data: { prefixClass },
      } = getCurrentInstance();
      const { footerAlign } = props;
      let footerClass = [`${prefixClass}-modal--footer`];

      if (footerAlign) footerClass.push(`align-${footerAlign}`);

      return footerClass;
    });

    onMounted(() => {
      if (
        Reflect.has(modalContext, 'pushModal') &&
        modalContext !== undefined
      ) {
        modalContext.pushModal(root.vnode);
      }

      isShow.value = props.onShow;
    });
    return {
      modal,
      modalEl,
      modalBodyEl,
      modalFooterEl,
      modalConfirmBtnEl,
      modalCancelBtnEl,
      isShow,
      handleConfirm,
      handleCancel,
      toggleModal,
      getContent,
      getFooter,
      getFooterClass,
    };
  },
};
</script>
<style lang="scss" scoped>
.#{$prefixClass}-modal {
  @apply fixed z-50 inset-0 overflow-y-auto;

  &--container {
    @apply flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0;
  }

  &--overlay {
    @apply fixed inset-0 transition-opacity;

    &-bg {
      @apply absolute inset-0 bg-gray-500 opacity-75;
    }
  }

  &--body {
    @apply inline-block align-bottom bg-white rounded-lg text-left z-50;
    @apply overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-lg sm:w-full;
  }

  &--footer {
    @apply bg-gray-50 pb-5 px-5 flex flex-row-reverse;

    &.align-left {
      @apply justify-end;
    }

    &.align-center {
      @apply justify-center;
    }
  }

  &--panel {
    @apply bg-white p-5;

    &-content {
      @apply mt-3 text-center sm:mt-0 sm:text-left;

      .content-header {
        @apply mb-4 text-lg leading-6 font-medium text-gray-900;
      }

      .content-desc {
      }
    }
  }
}
</style>
