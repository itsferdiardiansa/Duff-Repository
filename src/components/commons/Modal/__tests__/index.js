import { mount } from '@vue/test-utils'
import Modal from '@common/Modal'

describe('common/Modal', () => {
  let Component
  let prefixClass
  let handleOnConfirm = jest.fn()
  let handleOnCancel = jest.fn()

  beforeAll(() => {
    Component = mount(Modal, {
      slots: {
        header: 'Delete confirmation',
        body: 'Are you sure to delete this item?',
      },
    })
    prefixClass = `${Component.rootVM.prefixClass}-modal`
  })

  beforeEach(() => {
    Component = mount(Modal)
  })

  it('will rendered with snapshot', async () => {
    await self.$modal.show()

    expect(Component.element).toMatchSnapshot()
  })

  it('will rendered with passing title', async () => {
    await Component.setProps({ title: 'This is title' })

    await self.$modal.show()

    expect(Component.html()).toContain('This is title')
    expect(Component.element).toMatchSnapshot()
  })

  it('will rendered with passing a descripotio', async () => {
    await Component.setProps({ description: 'This is a description of modal' })

    await self.$modal.show()

    expect(Component.html()).toContain('This is a description of modal')
    expect(Component.element).toMatchSnapshot()
  })

  it('will disappear after the user clicks the confirmation button', async () => {
    await Component.setProps({
      title: 'Delete',
      description: 'Are you sure to delete this item?',
      onConfirmFn: handleOnConfirm,
    })

    await self.$modal.show()

    // Click to confirm
    await Component.vm.modalFooterEl.querySelector('button').click()

    // Modal will be hide from DOM
    expect(Component.vm.modalEl).toBeNull()
  })
})
