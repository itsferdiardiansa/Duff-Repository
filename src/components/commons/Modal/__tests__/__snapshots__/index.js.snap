// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Modal will rendered with passing a descripotio 1`] = `
<div
  class="lk-modal"
>
  <div
    class="lk-modal--container"
  >
    <div
      aria-hidden="true"
      class="lk-modal--overlay"
    >
      <div
        class="lk-modal--overlay-bg"
      />
    </div>
    <span
      class="hidden sm:inline-block sm:align-middle sm:h-screen"
    >
      ​
    </span>
    <div
      class="lk-modal--body fade-slide-enter-from fade-slide-enter-active"
      role="dialog"
    >
      <div
        class="lk-modal--panel"
      >
        <div
          class="lk-modal--panel-content"
        >
          <h3
            class="content-header"
          >
            
            
            
          </h3>
          <div
            class="content-desc"
          >
            
            This is a description of modal
            
          </div>
        </div>
      </div>
      <div
        class="lk-modal--footer"
      >
        <button
          class="lk-button lk-button--success lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              Yes
              
            </div>
          </div>
        </button>
        <button
          class="lk-button lk-button--danger lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              No
              
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`common/Modal will rendered with passing title 1`] = `
<div
  class="lk-modal"
>
  <div
    class="lk-modal--container"
  >
    <div
      aria-hidden="true"
      class="lk-modal--overlay"
    >
      <div
        class="lk-modal--overlay-bg"
      />
    </div>
    <span
      class="hidden sm:inline-block sm:align-middle sm:h-screen"
    >
      ​
    </span>
    <div
      class="lk-modal--body fade-slide-enter-from fade-slide-enter-active"
      role="dialog"
    >
      <div
        class="lk-modal--panel"
      >
        <div
          class="lk-modal--panel-content"
        >
          <h3
            class="content-header"
          >
            
            This is title
            
          </h3>
          <div
            class="content-desc"
          >
            
            
            
          </div>
        </div>
      </div>
      <div
        class="lk-modal--footer"
      >
        <button
          class="lk-button lk-button--success lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              Yes
              
            </div>
          </div>
        </button>
        <button
          class="lk-button lk-button--danger lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              No
              
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`common/Modal will rendered with snapshot 1`] = `
<div
  class="lk-modal"
>
  <div
    class="lk-modal--container"
  >
    <div
      aria-hidden="true"
      class="lk-modal--overlay"
    >
      <div
        class="lk-modal--overlay-bg"
      />
    </div>
    <span
      class="hidden sm:inline-block sm:align-middle sm:h-screen"
    >
      ​
    </span>
    <div
      class="lk-modal--body fade-slide-enter-from fade-slide-enter-active"
      role="dialog"
    >
      <div
        class="lk-modal--panel"
      >
        <div
          class="lk-modal--panel-content"
        >
          <h3
            class="content-header"
          >
            
            
            
          </h3>
          <div
            class="content-desc"
          >
            
            
            
          </div>
        </div>
      </div>
      <div
        class="lk-modal--footer"
      >
        <button
          class="lk-button lk-button--success lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              Yes
              
            </div>
          </div>
        </button>
        <button
          class="lk-button lk-button--danger lk-button--base text-bold"
        >
          <div
            class="lk-button--wrapper"
          >
            <div
              class="lk-button--icon"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="lk-button--label has-icon"
            >
              
              No
              
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
